core/state/snapshot/iterator_binary.go

```diff
@@ -117,33 +117,46 @@ func (dl *diffLayer) initBinaryStorageIterator(account common.Hash) Iterator {
 // or an error if iteration failed for some reason (e.g. root being iterated
 // becomes stale and garbage collected).
 func (it *binaryIterator) Next() bool {
+	for {
+		if !it.next() {
+			return false
+		}
+		if len(it.Account()) == 0 && len(it.Slot()) == 0 {
```

Comment:
We should also check the it.Fail. 

`it.Fail` might be set if any error occurs by calling `it.Account()` or `it.Slot()`. If so, we should return false and stop iteration.